File: src/libc/string.c
#include <string.h>


// K&R implementation
void itoa (uint32_t n, char s[]) 
{
  uint32_t i, sign;
  if ((sign = n) < 0) 
    n = -n;
  i = 0;
  do
  {
    s[i++] = n % 10 + '0';
  } while ((n /= 10) > 0);

  if (sign < 0)
    s[i++] = '-';
  s[i] = '\0';

  reverse(s);
}

// K&R
void reverse(char s[]) 
{
  int c, i, j;
  for (i = 0, j = strlen(s)-1; i < j; i++, j--)
  {
    c = s[i];
    s[i] = s[j];
    s[j] = c;
  }
}

// K&R
int strlen (char s[]) 
{
  int i = 0;
  while (s[i] != '\0')
    ++i;
  return i;
}

File: src/tests/libc/string.py
import gdb
import os
import sys
sys.path.append(os.path.dirname('..'))
import james_utils

utils = james_utils.JamesUtils()
utils.GdbInit()

## TEST string.h

### KHEAP_INIT
gdb.execute('call kheap_init()')
string = gdb.parse_and_eval('kmalloc_u(10)')

### ITOA
### CANNOT TEST SIGNED CONVERSION, variables are unsigned
gdb.execute(f'call itoa(12345678, {string})')
expected = b'12345678'
assert(utils.TestMemoryRegion(expected, string))
print(f'\033[92m[+] STRING itoa(12345678, {string}) executado corretamente.\033[0m')

### REVERSE
gdb.execute(f'call reverse({string})')
expected = b'87654321'
assert(utils.TestMemoryRegion(expected, string))
print(f'\033[92m[+] STRING reverse({string}) executado corretamente.\033[0m')

### STRLEN
size = gdb.parse_and_eval(f'strlen({string})')
assert(size == 8)
print(f'\033[92m[+] STRING reverse({string}) executado corretamente.\033[0m')


utils.GdbStop()
